### PR TITLE
Quest type: hunting location

### PR DESCRIPTION
**Pull request by Balferian**
* **Addressed Issue(s)**: https://github.com/rathena/rathena/issues/5490
* **Server Mode**: Pre/Re
* **Description of Pull Request**: 
Add support for new quest type: hunting location.

### CODE REVIEW

File: src/map/quest.hpp [hunk 1]
```diff
@@ -33,6 +33,8 @@ struct s_quest_objective {
 	e_race race;
 	e_size size;
 	e_element element;
+	std::string location;
+	std::string map_name;
```
Comment (Lemongrass3110) [inline src/map/quest.hpp:37]: Is the mapname different from the "normal" mapname or can the existing name be used?
If it is the same I would not store it here.

File: src/map/quest.hpp [hunk 2]
```diff
@@ -33,6 +33,8 @@ struct s_quest_objective {
 	e_race race;
 	e_size size;
 	e_element element;
+	std::string location;
```
Comment (Lemongrass3110) [inline src/map/quest.hpp:36]: Please store the map by mapid for faster checking.

File: src/map/quest.hpp (hunk 1 above)
Comment (Balferian) [reply src/map/quest.hpp:37]: The "normal" name of the map is used like in mapInfo.lub, I have not seen another name.

File: src/map/clif.cpp [hunk 3]
```diff
@@ -17164,7 +17164,7 @@ static void clif_quest_len(int def_len, int info_len, int avail_quests, int *lim
 	(*len_out) = ((*limit_out) * info_len) + def_len;
 }
 
-std::string clif_mobtype_name(e_race race, e_size size, e_element element) {
+std::string clif_mobtype_name(e_race race, e_size size, e_element element, std::string map_name) {
```
Comment (lighta) [inline src/map/clif.cpp:17167]: `const std::string& map_name)
no need to copy this

Comment (Lemongrass3110) [inline src/map/clif.cpp:17167]:
```suggestion
std::string clif_mobtype_name(e_race race, e_size size, e_element element, std::string& map_name) {
```

File: src/map/clif.cpp [hunk 4]
```diff
@@ -17257,6 +17258,7 @@ void clif_quest_send_list(struct map_session_data *sd)
 				e_race race = qi->objectives[j]->race;
 				e_size size = qi->objectives[j]->size;
 				e_element element = qi->objectives[j]->element;
+				std::string map_name = qi->objectives[j]->map_name;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:17261]:
```suggestion
```

File: src/map/clif.cpp [hunk 5]
```diff
@@ -17279,7 +17281,7 @@ void clif_quest_send_list(struct map_session_data *sd)
 				if (mob && qi->objectives[j]->mob_id > 0)
 					safestrncpy((char *)WFIFOP(fd,offset), mob->jname, NAME_LENGTH);
 				else
-					safestrncpy((char *)WFIFOP(fd,offset), clif_mobtype_name(race, size, element).c_str(), NAME_LENGTH);
+					safestrncpy((char *)WFIFOP(fd,offset), clif_mobtype_name(race, size, element, map_name).c_str(), NAME_LENGTH);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:17284]:
```suggestion
					safestrncpy(WFIFOCP(fd,offset), clif_mobtype_name(race, size, element, qi->objectives[j]->map_name).c_str(), NAME_LENGTH);
```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -17377,6 +17379,7 @@ void clif_quest_add(struct map_session_data *sd, struct quest *qd)
 		e_race race = qi->objectives[i]->race;
 		e_size size = qi->objectives[i]->size;
 		e_element element = qi->objectives[i]->element;
+		std::string map_name = qi->objectives[i]->map_name;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:17382]:
```suggestion
```

File: src/map/clif.cpp [hunk 7]
```diff
@@ -17397,7 +17400,7 @@ void clif_quest_add(struct map_session_data *sd, struct quest *qd)
 		if (mob && qi->objectives[i]->mob_id > 0)
 			safestrncpy((char *)WFIFOP(fd,offset), mob->jname, NAME_LENGTH);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:17401]:
```suggestion
			safestrncpy(WFIFOCP(fd,offset), mob->jname, NAME_LENGTH);
```

File: src/map/clif.cpp [hunk 8]
```diff
@@ -17397,7 +17400,7 @@ void clif_quest_add(struct map_session_data *sd, struct quest *qd)
...
 		else
-			safestrncpy((char *)WFIFOP(fd,offset), clif_mobtype_name(race, size, element).c_str(), NAME_LENGTH);
+			safestrncpy((char *)WFIFOP(fd,offset), clif_mobtype_name(race, size, element, map_name).c_str(), NAME_LENGTH);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:17403]:
```suggestion
			safestrncpy(WFIFOCP(fd,offset), clif_mobtype_name(race, size, element, qi->objectives[i]->map_name).c_str(), NAME_LENGTH);
```

File: src/map/quest.cpp [hunk 9]
```diff
@@ -163,7 +163,7 @@ uint64 QuestDatabase::parseBodyNode(const YAML::Node &node) {
 				}
 
 				if (!this->nodeExists(targetNode, "Mob") && !this->nodeExists(targetNode, "MinLevel") && !this->nodeExists(targetNode, "MaxLevel") &&
-						!this->nodeExists(targetNode, "Race") && !this->nodeExists(targetNode, "Size") && !this->nodeExists(targetNode, "Element")) {
+						!this->nodeExists(targetNode, "Race") && !this->nodeExists(targetNode, "Size") && !this->nodeExists(targetNode, "Element") && !this->nodeExists(targetNode, "MapName")) {
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:166]:
```suggestion
						!this->nodeExists(targetNode, "Race") && !this->nodeExists(targetNode, "Size") && !this->nodeExists(targetNode, "Element") &&
						!this->nodeExists(targetNode, "Location") && !this->nodeExists(targetNode, "MapName")) {
```

File: src/map/quest.cpp [hunk 10]
```diff
@@ -268,6 +270,38 @@ uint64 QuestDatabase::parseBodyNode(const YAML::Node &node) {
...
 
+				if (this->nodeExists(targetNode, "Location")) {
+					std::string location;
+
+					if (!this->asString(targetNode, "Location", location))
+						return 0;
+
+					unsigned short mapindex;
+					int16 m = -1;
+					mapindex = mapindex_name2idx(location.c_str(), nullptr);
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:281]:
```suggestion
					uint16 mapindex = mapindex_name2idx(location.c_str(), nullptr);

```

File: src/map/quest.cpp [hunk 11]
```diff
@@ -176,6 +176,8 @@ uint64 QuestDatabase::parseBodyNode(const YAML::Node &node) {
 				target->race = RC_ALL;
 				target->size = SZ_ALL;
 				target->element = ELE_ALL;
+				target->location = "";
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:179]:
```suggestion
				target->mapid = -1;
```

File: src/map/quest.cpp [hunk 12]
```diff
@@ -268,6 +270,38 @@ uint64 QuestDatabase::parseBodyNode(const YAML::Node &node) {
...
+					if (!this->asString(targetNode, "Location", location))
+						return 0;
+
+					unsigned short mapindex;
+					int16 m = -1;
+					mapindex = mapindex_name2idx(location.c_str(), nullptr);
+					if (mapindex)
+						m = map_mapindex2mapid(mapindex);
+
+					if (m < 0) {
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:285]:
```suggestion
					if (mapindex == 0) {
```

File: src/map/quest.cpp [hunk 13]
```diff
@@ -268,6 +270,38 @@ uint64 QuestDatabase::parseBodyNode(const YAML::Node &node) {
...
+						this->invalidWarning(targetNode["Location"], "Map \"%s\" not found.\n", location.c_str());
+						return 0;
+					}
+
+					target->location = location;
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:290]:
```suggestion
					target->mapid = map_mapindex2mapid(mapindex);
```

File: src/map/quest.cpp [hunk 14]
```diff
@@ -268,6 +270,38 @@ uint64 QuestDatabase::parseBodyNode(const YAML::Node &node) {
...
+				}
+
+				if (this->nodeExists(targetNode, "MapName")) {
+					std::string map_name;
+
+					if (target->location == "")
+						return 0;
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:297]:
```suggestion
```

File: src/map/quest.cpp [hunk 15]
```diff
@@ -708,9 +742,11 @@ void quest_update_objective(struct map_session_data *sd, int mob_id, int mob_lev
 					objective_check++;
 				if (qi->objectives[j]->element == ELE_ALL || qi->objectives[j]->element == mob_element)
 					objective_check++;
+				if (qi->objectives[j]->location == "" || qi->objectives[j]->location == mapindex_id2name(sd->mapindex))
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:745]:
```suggestion
				if (qi->objectives[j]->mapid < 0 || qi->objectives[j]->mapid== sd->bl.m)
```

File: src/map/quest.hpp (hunk 2 above)
Comment (Lemongrass3110) [inline src/map/quest.hpp:36]:
```suggestion
	int16 mapid;
```

**Review by Lemongrass3110 — CHANGES_REQUESTED** [review]

**Review by Lemongrass3110 — APPROVED** [review]

**Review by aleos89 — APPROVED** [review]

**Pull request merged**